exp/ipam/internal/webhooks/ipaddress.go

```diff
@@ -0,0 +1,169 @@
+// +kubebuilder:rbac:groups=ipam.cluster.x-k8s.io,resources=ipaddressclaims,verbs=get;list;watch
+
+// IPAddress implements a validating webhook for IPAddress.
+type IPAddress struct {
+	Client client.Reader
+}
+
+var _ webhook.CustomValidator = &IPAddress{}
+
+// ValidateCreate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	ip, ok := obj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", obj))
+	}
+	return webhook.validate(ctx, ip)
+}
+
+// ValidateUpdate implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	oldIP, ok := oldObj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", oldObj))
+	}
+	newIP, ok := newObj.(*ipamv1.IPAddress)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected an IPAddress but got a %T", newObj))
+	}
+
+	if !reflect.DeepEqual(oldIP.Spec, newIP.Spec) {
+		return field.Forbidden(field.NewPath("spec"), "the spec of IPAddress is immutable")
+	}
+	return nil
+}
+
+// ValidateDelete implements webhook.CustomValidator.
+func (webhook *IPAddress) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return nil
+}
+
+func (webhook *IPAddress) validate(ctx context.Context, ip *ipamv1.IPAddress) error {
+	log := ctrl.LoggerFrom(ctx)
+	allErrs := field.ErrorList{}
+
+	addr, err := netaddr.ParseIP(ip.Spec.Address)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.address"),
+				ip.Spec.Address,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.Prefix < 0 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix cannot be negative",
+			))
+	}
+	if addr.Is4() && ip.Spec.Prefix > 32 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv4 address",
+			))
+	}
+	if addr.Is6() && ip.Spec.Prefix > 128 {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.prefix"),
+				ip.Spec.Prefix,
+				"prefix is too large for an IPv6 address",
+			))
+	}
+
+	_, err = netaddr.ParseIP(ip.Spec.Gateway)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.gateway"),
+				ip.Spec.Gateway,
+				"not a valid IP address",
+			))
+	}
+
+	if ip.Spec.PoolRef.APIGroup == nil {
+		allErrs = append(allErrs,
+			field.Invalid(
+				field.NewPath("spec.poolRef.apiGroup"),
+				ip.Spec.PoolRef.APIGroup,
+				"the pool reference needs to contain a group"))
+	}
+
+	claim := &ipamv1.IPAddressClaim{}
+	err = webhook.Client.Get(ctx, types.NamespacedName{Name: ip.Spec.ClaimRef.Name, Namespace: ip.ObjectMeta.Namespace}, claim)
```

Comment:
I was under the impression that we agree to drop this call to avoid the introduction of order in creating things that can impact clusterctl move or valero backup restore, am I wrona g?
On similar note, are there some ideas about move / owerRef chain?

Reply:
The current implementation is best effort. I.e. if the claim is there we validate it otherwise we skip over the check

Reply:
kk. what about move / ownerRef chain?

Reply:
Is that validated on other resources currently?

Reply:
I think Fabrizio's comment is not about validation.

It's more a general comment how would `clusterctl move` work with IPAddress and IPAddressClaim. For other resources it usually works (afaik) through discoverying resources via ownerRefs.

Reply:
@schrej  I took a look at this with @sbueringer and my understanding is the following

- ProviderIPPool CRD must have the `clusterctl.cluster.x-k8s.io/move` label, because they are not related to any cluster
- IPAddressClaim should have an ownerRef to the corresponding infraMachine; this will ensure move works given that infraMachine chain is already linked to a cluster
 - IPAddress should have an ownerRef to the IPAddressClaim that originated it; this will ensure move works given that IPAddressClaim ownerRef chain is already linked to a cluster
 
Does this sound ok for you? 
 
Something else that should be considered is what happens during move.
- First IPPool will be created in the target env before the IPAddessClaim, and then Claim will be created before IPAddess
- At this stage, for a brief period, those objects exist in two envs
- Then IPAddress, IPClaim and IPPools will be deleted from the original environment, please note that before deletion move removes the finalizer

What is important is that this operation should not trigger actual deletion of IPPool/release of IPAddress in the IPAM provider. Do you see problems in implementing this? We have some idea but we should probably talk about this

However, our feeling is that if you agree on the suggested ownerRef, we can merge this PR and defer everything to follow up changes down the line

Reply:
It does sound ok.
[Owner refs on Claims](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md?plain=1#L367) are already in the proposal.
Owner refs from address to claim should probably be added there explicitly, same for that move label.

We should definitely talk about move behaviour, as that might be tricky to implement. I'm not familiar with the internals of that process, especially if/how controllers can detect that a move is going on.

I'd say we can move forward with this PR then.